Add actions to run OpenLane

vvbandeira opened the pull request:
Usage is shown here: https://github.com/The-OpenROAD-Project/actions-test/actions
These actions allow users to test their designs with a stable and the latest version of OpenLane as well as the latest version of OpenROAD.

Signed-off-by: Vitor Bandeira <vitor.vbandeira@gmail.com>

vvbandeira committed:
Add actions

Signed-off-by: Vitor Bandeira <vitor.vbandeira@gmail.com>

tspyrou commented:
@vvbandeira Can you add a link to the example repo which uses the actions as well?

vvbandeira commented:
> @vvbandeira Can you add a link to the example repo which uses the actions as well?

It was as a link on the PR description, but I made more clear.

tspyrou commented:
I also added @proppy and asked him to review on behalf of google.

proppy commented on cache_pdks/action.yml:
```diff
@@ -0,0 +1,49 @@
+inputs:
+  pdk_root:
+    description: PDK root to install files.
+    requried: true
```
required?

vvbandeira replied:
Fixed on [29f5492](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/29f5492b82fd1ee0091f8cc0586d7d96981877dc)

proppy commented on openlane_pr_tag/action.yml:
```diff
@@ -0,0 +1,34 @@
+          See the individual commits for details.
+        commit-message: |
+          [BOT] Update OpenLane stable tag -> ${{ inputs.ol_tag }}
+        branch: openlane-tag-update-branch
```
shouldn't we include the tag inside the branch name?

vvbandeira replied:
I think this is a decision we need to make. If we use the tag on the branch name, then each new tag (i.e., each time the action is run) would be a new PR. Keeping the tag name out of the branch name the action would update an open PR.

proppy replied:
Make senses, I didn't realize that the actions runs would need to stack up.

proppy commented on openlane_pr_tag/action.yml:
```diff
@@ -0,0 +1,34 @@
+      shell: bash
+      run: |
+        cd ${{ inputs.path }}
+        echo "${{ inputs.ol_tag }}" > .github/openlane-stable-tag
```
is this file used by other actions?

vvbandeira replied:
Yes, there's an example on the intended use [here](https://github.com/The-OpenROAD-Project/actions-test/blob/main/.github/workflows/openlane-stable.yml). But I have also changed it to be able to input on [df1d269](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/df1d269fa71afb02518ae0df9efabbf67ba89857).

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,31 @@
+    - name: Export PDK ROOT
+      shell: bash
+      run: |
+        echo "PDK_ROOT="${{ inputs.pdk_root }} >> $GITHUB_ENV
```
giving that those are directly derived from the `inputs`, I'm curious if we could define those as regular `env` keys https://docs.github.com/en/actions/learn-github-actions/environment-variables#about-environment-variables here?

vvbandeira replied:
I was just following the "style" on the repo and other actions on the organization. But I will take look at if env can resolve a `${{ input.xyz }}` statement.

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,37 @@
+
+    - name: Update tool hash
+      shell: bash
+      run: python3 .github/scripts/update_tools.py ${{ inputs.tool_name }}
```
isn't the `update_tools.py` script tied to a particular repo?

vvbandeira replied:
Yes, for OpenLane. This action is intended to be used for workflows that use that repo.

proppy replied:
if the action is tied to openlane, shouldn't it live in https://github.com/The-OpenROAD-Project/OpenLane/tree/master/.github/workflows ?

vvbandeira replied:
The goal of these actions is to allow the orchestration of the test and update OL version without touching OL repo. We might consider moving the action later if OL devs agree.

proppy replied:
oh I see, so this is the local equivalent of https://github.com/The-OpenROAD-Project/OpenLane/blob/master/.github/workflows/tool_updater.yml, to specify that you want to update one of the tool to the latest version before running `openlane_run`?

vvbandeira replied:
Yes, that's the intended use. See [here](https://github.com/The-OpenROAD-Project/actions-test/blob/main/.github/workflows/openroad-latest.yml#L21-L25) for an example.

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,37 @@
+  build_ol:
+    description: Build docker image after tool update.
+    default: true
+
```
should there be an input tag as well?

vvbandeira replied:
Added on [d74b2bd](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/d74b2bd9e00a47caec055071ecd2733318b4c9ae)

proppy commented on cache_pdks/action.yml:
```diff
@@ -0,0 +1,49 @@
+  pdk_root:
+    description: PDK root to install files.
+    required: true
+  ol_path:
```
Does this actually need to be part of the inputs? Since openlane is just one way to get/build a particular open_pdks version, it doesn't necessarly need to bleed in the action contract.

vvbandeira replied:
The goal of these actions is for use with OpenLane, see also my comment above about magic version. For these reasons, I think it's ok to keep ol_path.

proppy replied:
oh I see, what's confusing to me is that it uses `ol_path` working copy to extract the `OPEN_PDKS_COMMIT` (which could be on any versions) and uses `ol_tag` for actually building the PDK in case of a cache miss.

Shouldn't we always make sure the two are matching (either my sticking with the version defined in `ol_path` or extracting the one from `ol_tag`)?

vvbandeira replied:
I have changed to use only `ol_tag`.

proppy commented on cache_pdks/action.yml:
```diff
@@ -0,0 +1,49 @@
+      if: steps.cache.outputs.cache-hit != 'true'
+      shell: bash
+      run: |
+        make -C ${{ inputs.ol_path }} \
```
any reason we need to use openlane here instead of simply building the open_pdks repo? (could be way faster since you wouldn't have to fetch the openlane container image)?

vvbandeira replied:
See above.

proppy commented on cache_pdks/action.yml:
```diff
@@ -0,0 +1,49 @@
+    description: OpenLane path.
+    default: .
+  ol_tag:
+    description: OpenLane docker tag.
```
Curious if we could make this more reusable (outside of an openlane context) by specifying the open_pdks version instead?

vvbandeira replied:
I could refactor and create a generic version of this action and then an OL specific wrapper. I wanted to keep this as straightforward as possible for OL users.

proppy replied:
see other comments, I think the potentially confusing part is is that there are potentially two version of openlane at play here (the one the workflow has checked-out in `ol_path` and the one specified in `ol_tag`).

Did you consider having the action do the openlane checkout by itself so that you can have clearer assumption about which version to use?

vvbandeira replied:
If this action also makes the checkout of OL then there would OL would be on two paths. I have renamed the action to `openlane_cache_pdk` and removed `ol_path`, thus requiring OL to be already checked out.

proppy commented on cache_pdks/action.yml:
```diff
@@ -0,0 +1,49 @@
+name: Install PDKs.
```
note we have https://anaconda.org/LiteX-Hub/open_pdks.sky130a/files packages here that can also make it easy provision arbitrary version of the pdks without having to build them (also see https://github.com/hdl/conda-eda/issues/203)

vvbandeira replied:
OpenLane complains if the PDKs were built with a different version of magic than the version they validated. See here: https://github.com/The-OpenROAD-Project/OpenLane/blob/7891c2c8859e8c800226920c181fa039ad298e79/dependencies/verify_versions.py#L143
Unless we can pick both the PDK version and the magic version I would rather keep as is to avoid warnings during build or maybe issues with incompatibility in the future.

proppy replied:
make sense, the recently introduced a way to selectively disable those check https://github.com/The-OpenROAD-Project/OpenLane/pull/1042 but I guess the intent here is to test against the versions blessed by OpenLane?

Conda does allow you to pick the open_pdks version and the magic version with version selectors, but I think that might be a wider topic outside of the scope of this PR (similar to the discussion we have w/ @donn about sharing release infrastructure in https://github.com/efabless/volare/issues/4)

vvbandeira replied:
Yes, the goal is to test with blessed versions. We can discuss and refactor later if we agree on a different aproach.

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,40 @@
+      shell: bash
+      run: python3 .github/scripts/update_tools.py ${{ inputs.tool_name }}
+
+    - name: Build OpenROAD docker image
```
alternatively we have pre-built packages for openroad https://anaconda.org/LiteX-Hub/openroad/files

vvbandeira replied:
Typo, should read "tool".

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,31 @@
+        echo "PDK_ROOT="${{ inputs.pdk_root }} >> $GITHUB_ENV
+        echo "OPENLANE_IMAGE_NAME=efabless/openlane:${{ inputs.ol_tag }}" >> $GITHUB_ENV
+
+    - name: Build OpenLane docker image
```
Does it actually build the OpenLane docker image or just fetch it?

vvbandeira replied:
Pull or build if not available.

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,40 @@
+      shell: bash
+      run: python3 .github/scripts/update_tools.py ${{ inputs.tool_name }}
+
+    - name: Build OpenROAD docker image
```
this also look more generic that just OpenROAD (as it can build any tool)

vvbandeira replied:
Yes, I originally wrote with update only OpenROAD in mind and then changed it to be generic. I missed this name, thanks. Fixed on 8358cfd

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,40 @@
+                       OPENLANE_DOCKER_TAG=${{ inputs.ol_tag }} \
+                       build-${{ inputs.tool_name }}
+
+    - name: Build OpenLane docker image with new OpenROAD bianry
```
s/bianry/binary.

vvbandeira replied:
Stick fingers. Fixed on 8358cfd

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,40 @@
+  tool_name:
+    description: Name of the tool to update.
+    required: true
+  pdk_root:
```
couldn't this just assume PDK_ROOT? assuming `cache_pdks` is always present as a siblings action in the workflow?

vvbandeira replied:
See above for required input. Also making it independent of other actions might be useful in the future.

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,31 @@
+  design_name:
+    description: Name of design to run
+    required: true
+  pdk_root:
```
couldn't this just assume PDK_ROOT? assuming cache_pdks is always present as a siblings action in the workflow?

vvbandeira replied:
Yes, but I think to require it as input is cleaner. We won't have to rely on the underlying tool to provide a meaningful error message or the user to know how to fix it.

proppy replied:
maybe it should default on `PDK_ROOT` then, or get it from `cache_pdks` inputs? (so that it's less burden on the users to have them always match)

proppy commented on openlane_pr_tag/action.yml:
```diff
@@ -0,0 +1,40 @@
+        echo "${{ inputs.ol_tag }}" > ${{ inputs.ol_tag_file }}
+
+    - name: Export tag name
+      run: echo "OL_TAG=${{ inputs.ol_tag }}" >> $GITHUB_ENV
```
is the only reason we're exporting here is to be able to reuse it in the commit message?

vvbandeira replied:
Yes, before adb37c90a203c5ae214bbd45a0f33172a33b9928 the input was not being correctly resolved.

proppy commented on openlane_pr_tag/action.yml:
```diff
@@ -0,0 +1,40 @@
+  ol_tag:
+    description: OpenLane tag to use in PR
+    required: true
+  path:
```
does this actually change?

vvbandeira replied:
It's a choice for the user. We could also make a fixed value.

proppy replied:
I assume this will more or less always be used in combination with `openlane_run` to update the tool to the corresponding version that was succesfully tested?

Did you consider merging the two action together, so that the end-users don't have to wire them manually together?

vvbandeira replied:
Merged with `openlane_run` and added an `bool` option to control PR creation.

proppy reviewed (commented):
another round of comments (that are mostly questions because I do miss some context on the project), feel free to disregard them if you don't think they are helpful/applicable.

vvbandeira reviewed (commented):
@proppy, thanks for your comments.

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,31 @@
+    required: true
+  ol_tag:
+    description: OpenLane tag
+    default: latest
```
should the default be the version stored in `openlane-stable-tag` ?

vvbandeira replied:
I think `latest` is easier to set as a default. The goal is that the user would have on their repo two workflows one that uses `openlane_run` with `latest` and the other with the stable tag.

proppy commented on README.md:
```diff
@@ -67,3 +67,37 @@ public pull request.
 
 Pulls the upstream repository into the local repository.
 
+## [`cache_pdks`](./cache_pdks)
```
can we add a brief description to the action file themselves as well?

vvbandeira replied:
Added.

proppy commented on README.md:
```diff
@@ -67,3 +67,37 @@ public pull request.
+
+Checks if there is a cache for a given PDK install. The cache hash used
+consists of `open_pdks`, `sky130` and `magic` commit hashes. If it can not
+find a cached install, the cache is built. Arguments:
```
> the cache is built?

a new PDK is built according to the specified `ol_path` and `ol_tag`.

vvbandeira replied:
Changed

proppy commented on README.md:
```diff
@@ -67,3 +67,37 @@ public pull request.
+consists of `open_pdks`, `sky130` and `magic` commit hashes. If it can not
+find a cached install, the cache is built. Arguments:
+-   `pdk_root` [required]: where to install PDKs
+-   `ol_path` [optional, default=`.`]: path to OpenLane repo.
```
maybe mention that the OpenLane checkout step is a pre-requisite.

proppy commented on README.md:
```diff
@@ -67,3 +67,37 @@ public pull request.
+Run OpenLane `test` on a given design. Arguments:
+-   `design_name` [required]: name of design to run. The design should be
+    available at the `./desigins` folder and have a `config.tcl` file.
+-   `pdk_root` [required]: PDK root path.
```
maybe mention that using the `cache_pdks` action is a pre-requisite?

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,40 @@
+    default: .
+  build_ol:
+    description: Build docker image after tool update.
+    default: true
```
what happens if user don't set this to true, does `openlane_run` end up using the updated version?

vvbandeira replied:
If the user wants to update multiple tools in one workflow they can skip the build on all but the last update. Added note about the usage on the docs.

proppy replied:
so in all cases the last `openlane_update_tool` of any workflow will always needs to have `build_ol=true` for `openlane_run` to benefit from it?

did you consider taking a list of tools to update instead? (github action don't support list but they apparently support multiline strings)

proppy commented on README.md:
```diff
@@ -67,3 +67,37 @@ public pull request.
 
 Pulls the upstream repository into the local repository.
 
+## [`cache_pdks`](./cache_pdks)
```
maybe rename to `openlane_cache_pdks` to match the other actions.

vvbandeira replied:
Renamed

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,40 @@
+      if: inputs.build_ol == 'true'
+      shell: bash
+      run: |
+        make -C docker PDK_ROOT=${{ inputs.pdk_path }} \
```
how do you ensure that this is getting picked up by `openlane_run`? just the fact that the user will have the `tag`s match?

vvbandeira replied:
`PDK_ROOT` here AFAIK is not very critical. The main reason to set is that OL fails otherwise.

proppy replied:
sorry, my comment was about the image being built here, if a user want's to do `openlane_run` against the version being built here, do they have to make sure they use the same `ol_tag` in both place or should the fact that a preceding step use this action be enough? (assuming the default matches or there is a way to hint that this workflow was run)

proppy commented:
More general feedbacks.

It seems that the current implementation assumes that
- the actions are run in a particular order (`cache_pdks`, `openlane_update_tool` optionally, `openlane_run` and eventually `openlane_pr_tag`)
- the user will checkout the openlane repo as a pre-requisite
- the user will wire all actions parameters to have matching `pdk_root`, `ol_tag` and `ol_path`.

I'm curious if you considered a more monolitic approach where the user specify a single step with option to optionally override a particular OpenLane version and update the tag stamp in `.github`.

proppy commented:
> I'm curious if you considered a more monolitic approach where the user specify a single step with option to optionally override a particular OpenLane version and update the tag stamp in .github.

One example of integration with such a monolitich action can be found here:
https://github.com/olofk/serv/blob/main/.github/workflows/openlane.yml#L9

If you want to keep the discrete steps separate, you could still consider providing a wrapper `composite` action that combine them in the most common scenario, so that it could be used with?
```
    steps:
      - name: Checkout design
        uses: actions/checkout@v2
      - name: OpenLane latest
        uses: The-OpenROAD-Project/openlane_run_flow
        with:
          config: path/to/my/openlane/config 
```
wdyt?

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,90 @@
+      If the test is successful update stable tag pointed by the ol_tag_file
+      input.
+    default: false
+  ol_tag_file:
```
shouldn't there be a way to run the design against the version pointed at by `ol_tag_file`?

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,90 @@
+      uses: vvbandeira/actions/openlane_cache_pdks@new
+      with:
+        pdk_root: ${{ inputs.pdk_root }}
+        ol_tag: ${{ inputs.ol_tag }}
```
should this use `OPENLANE_STABLE_TAG`?

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,90 @@
+      shell: bash
+      run: |
+        echo "PDK_ROOT="${{ inputs.pdk_root }} >> $GITHUB_ENV
+        if [[ -z "${OPENLANE_STABLE_TAG+x}" ]]; then
```
are you sure you need to do `+x` here?

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,90 @@
+        pdk_root: ${{ inputs.pdk_root }}
+        ol_tag: ${{ inputs.ol_tag }}
+
+    - name: Export PDK ROOT
```
this technically also export `OPENLANE_DOCKER_TAG` and `OPENLANE_DOCKER_TAG`

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,90 @@
+        if [[ -z "${OPENLANE_STABLE_TAG+x}" ]]; then
+          echo "OPENLANE_DOCKER_TAG=${{ inputs.ol_tag }}" >> $GITHUB_ENV
+        else
+          echo "OPENLANE_DOCKER_TAG=${OPENLANE_STABLE_TAG}" >> $GITHUB_ENV
```
I wonder if this need a more general name, as it's being used later in the action outside of the context of docker.

vvbandeira replied:
I am exporting with this name as it is used on the OpenLane makefile. Should I use a more general name and assign the general name to OPENLANE_DOCKER_TAG when calling OpenLane makefile?

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,90 @@
+      run: echo "${OPENLANE_DOCKER_TAG}" > ${{ inputs.ol_tag_file }}
+
+    - name: Create Pull Request to update stable OpenLane tag
+      if: inputs.update_tag != 'false'
```
should we also test if the tag actually changed? (otherwise we might be creating empty PRs)

proppy commented on openlane_update_tool/action.yml:
```diff
@@ -0,0 +1,45 @@
+    - name: Build docker image for updated tool
+      shell: bash
+      run: |
+        make -C docker PDK_ROOT=${{ inputs.pdk_path }} \
```
do you know if openlane images are built with `BUILDKIT_INLINE_CACHE=1`, you should be able to benefit from the cache using https://docs.docker.com/engine/reference/commandline/build/#specifying-external-cache-sources if that's the case

vvbandeira replied:
They do not, at least `grep BUILDKIT_INLINE_CACHE` returns nothing.

proppy reviewed (commented):
any thoughts on my earlier high-level comments?

vvbandeira commented:
> any thoughts on my earlier high-level comments?

@proppy 
About monolithic, yes. I was implementing the approach. I still need to test, but I have added the code here in case you want to take a first look.

vvbandeira commented:
Moved to monolithic approach and updated to use volare to get PDKs

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,120 @@
+    - shell: bash
+      run: |
+        echo '::group::Get PDKs'
+        make pdk
```
q: does this use volare?

vvbandeira replied:
Yes, see [here](https://github.com/The-OpenROAD-Project/OpenLane/blob/master/Makefile#L106-L108).

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,120 @@
+        delete-branch: true
+
+    - shell: bash
+      if: ${{ steps.cpr.outputs.pull-request-number }}
```
q: does `create-pull-request` already log something?

vvbandeira replied:
It does, but is not as useful as the info on this step. This step was copied from their docs.
```
Pushing pull request branch to 'origin/update-openlane-tag'
  /usr/bin/git push --force-with-lease origin HEAD:refs/heads/update-openlane-tag
  remote: 
  remote: Create a pull request for 'update-openlane-tag' on GitHub by visiting:        
  remote:      https://github.com/The-OpenROAD-Project/actions-test/pull/new/update-openlane-tag        
  remote: 
  To https://github.com/The-OpenROAD-Project/actions-test
   * [new branch]      HEAD -> update-openlane-tag
Create or update the pull request
  Attempting creation of pull request
  Created pull request #2 (The-OpenROAD-Project:update-openlane-tag => main)
```
vs
```
Run echo "Pull Request Number - 2"
Pull Request Number - 2
Pull Request URL - https://github.com/The-OpenROAD-Project/actions-test/pull/2
```

proppy commented on README.md:
```diff
@@ -67,3 +67,17 @@ public pull request.
+using a given tag. Optionally a PR can be created to update a stable tag
+file stored in the repository.
+Arguments:
+-   `ol_tag` [optional, default: `latest`]: OpenLane tag. You can set to
```
shouldn't the default be `master` also curious of what you would think of setting it to `ol_tag_file`?

vvbandeira replied:
Fixed on [0b11b71](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/0b11b7106247798cb27386930a42419f6d632d64)

proppy commented on README.md:
```diff
@@ -67,3 +67,17 @@ public pull request.
+Arguments:
+-   `ol_tag` [optional, default: `latest`]: OpenLane tag. You can set to
+    `ol_tag_file` to use the value from the input.
+-   `update_tag` [optional, default: `false`]: if the test is successful
```
assuming this is set to `true`, wouldn't the tag get effectively updated only if `ol_tag` != `ol_tag_file`? And if that's the case wouldn't that defeat the purpose of the file itself (as it wouldn't get used as an input).

Or is the intent here that you'd have two interdependant workflow (one with `ol_tag: master` and `update_tag: true`, and another with `ol_tag: ol_tag_file` and `update_tag: false`)?

If that's the case maybe I'm curious if there is a better way to encode/documentation this pattern.

vvbandeira replied:
The intent is to have interdependent workflows as you said:
- Test any changes made to the design with the last known good version (i.e., `ol_tag_file`).
- Test each new release of OpenLane with the user's design, if it passes give the choice to update the latest tag to be the stable tag (i.e., master/HEAD -> `ol_tag_file`).

I would be ok with removing `ol_tag` and `update_tag` in favour of a `mode` with the following possible values: `stable` (uses the file, does not update the file), `latest` (uses master/HEAD, tries to update file). But this will reduce the flexibility of the action. I can think of at least one use case that would not be possible, e.g., users cannot create a dispatch workflow to test with arbitrary tags/patches. Note that I am not the intended demographic, so this is just my best effort to support them.

proppy replied:
q: maybe you could add a wrapper action that encaspulate this pattern? so that you can easily add it to your project.

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,120 @@
+        echo '::group::Update each tool hash'
+        for tool in ${{ inputs.tools_list }}
+        do
+          python3 .github/scripts/update_tools.py "${tool}"
```
q: reading thru https://github.com/The-OpenROAD-Project/OpenLane/blob/master/.github/scripts/update_tools.py it seems that this will get the latest version of the tools independently of what's specified in the `ol_tag`, can you confirm this is the intent?

vvbandeira replied:
Yes, the goal is to use the latest version independent of `ol_tag`.

proppy replied:
q: should we hightlight this in the documentation for that flag?

vvbandeira replied:
Added on [be164f1](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/be164f1d1b19c9f12659d6e2f130c912eebf2d05)

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,120 @@
+    - uses: actions/checkout@v3
+      with:
+        repository: The-OpenROAD-Project/OpenLane
+        fetch-depth: 0
```
shouldn't this use `ol_tag` and `fetch-depth: 1`? (and maybe have a preceding step to resolve `ol_tag_file`)

vvbandeira replied:
We need the git history to make the PR. Is simpler to just clone the history.

proppy replied:
q: I guess the repo is small enough not to bother to doing a full clone every if you're not creating PR? (but you could in theory make the depth depend of the pr flag)

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,120 @@
+    default: .github/openlane-stable-tag
+  tools_list:
+    description: List of the tools to update separated by a space. Example "openroad_app magic".
+    default: None
```
should the default be an an empty string?

vvbandeira replied:
Fixed on [0b11b71](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/0b11b7106247798cb27386930a42419f6d632d64)

proppy commented on openlane_run/action.yml:
```diff
@@ -0,0 +1,120 @@
+        done
+        echo '::endgroup::'
+        echo '::group::Build OpenLane docker image with new tool binary'
+        make -C docker openlane
```
shouldn't we also do this when `ol_tag: master` even if the tool list is empty or can we assume that a container image has already been pushed?

vvbandeira replied:
Fixed on [0b11b71](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/0b11b7106247798cb27386930a42419f6d632d64)

proppy commented on README.md:
```diff
@@ -67,3 +67,17 @@ public pull request.
+    `ol_tag_file` to use the value from the input.
+-   `update_tag` [optional, default: `false`]: if the test is successful
+    update stable tag pointed by the `ol_tag_file` input.
+-   `ol_tag_file` [optional, default: `.github/openlane-stable-tag`]: File
```
would the action fail if the file doesn't exist or create a PR?

vvbandeira replied:
Fixed on [0b11b71](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/0b11b7106247798cb27386930a42419f6d632d64)

proppy reviewed (commented):
I really like the new direction, curious if you think it would be easier to use and maintain that way?

Some additional comments (mostly nits and questions)

vvbandeira committed:
PR reviews

latest -> master on docs
None -> ''

Always call get-openlane make target before running the design.
This target tries to pull existing docker image, otherwise it will build.

Signed-off-by: Vitor Bandeira <vitor.vbandeira@gmail.com>

vvbandeira commented:
@proppy were all your concerns addressed so we can merge this PR?

proppy commented on README.md:
```diff
@@ -67,3 +67,18 @@ public pull request.
+file stored in the repository.
+Arguments:
+-   `ol_tag` [optional, default: `master`]: OpenLane tag. You can set to
+    `ol_tag_file` to use the value from the input.
```
nit: "to use the value from the input" → "to use the value from the file content"?

vvbandeira replied:
Changed on [be164f1](https://github.com/The-OpenROAD-Project/actions/pull/11/commits/be164f1d1b19c9f12659d6e2f130c912eebf2d05)

proppy reviewed (approved):
Just a last pass of nits and qs, feel free to dismiss them if you don't find them relevant/applicable.

vvbandeira committed:
Docs improvments

Signed-off-by: Vitor Bandeira <vitor.vbandeira@gmail.com>
